The Taxi Data that is being worked with was downloaded from https://data.cityofchicago.org/Transportation/Taxi-Trips-2024-/ajtu-isnz/about_data.

The data we downloaded was from 01-01-2024 until 01-06-2026.
The dataset consists of 16.1 million rows and 23 columns. Each row is a trip of a taxi in Chigago, with *Trip ID* being its unique identifier.

This full description of the dataset is copied from the website.

| Column Name | Description | Data Type |
|-------------|----------------------------|-----------|
| Trip ID | A unique identifier for the trip. | Text |
| Taxi ID | A unique identifier for the taxi. | Text |
| Trip Start Timestamp | When the trip started, rounded to the nearest 15 minutes. | Floating Timestamp |
| Trip End Timestamp | When the trip ended, rounded to the nearest 15 minutes. | Floating Timestamp |
| Trip Seconds | Time of the trip in seconds. | Number |
| Trip Miles | Distance of the trip in miles. | Number |
| Pickup Census Tract | The Census Tract where the trip began. For privacy, this Census Tract is not shown for some trips. This column often will be blank for locations outside Chicago. | Text |
| Dropoff Census Tract | The Census Tract where the trip ended. For privacy, this Census Tract is not shown for some trips. This column often will be blank for locations outside Chicago. | Text |
| Pickup Community Area | The Community Area where the trip began. This column will be blank for locations outside Chicago. | Number |
| Dropoff Community Area | The Community Area where the trip ended. This column will be blank for locations outside Chicago. | Number |
| Fare | The fare for the trip. | Number |
| Tips | The tip for the trip. Cash tips generally will not be recorded. | Number |
| Tolls | The tolls for the trip. | Number |
| Extras | Extra charges for the trip. | Number |
| Trip Total | Total cost of the trip, the total of the previous columns. | Number |
| Payment Type | Type of payment for the trip. | Text |
| Company | The taxi company. | Text |
| Pickup Centroid Latitude | The latitude of the center of the pickup census tract or the community area if the census tract has been hidden for privacy. This column often will be blank for locations outside Chicago. | Number |
| Pickup Centroid Longitude | The longitude of the center of the pickup census tract or the community area if the census tract has been hidden for privacy. This column often will be blank for locations outside Chicago. | Number |
| Pickup Centroid Location | The location of the center of the pickup census tract or the community area if the census tract has been hidden for privacy. This column often will be blank for locations outside Chicago. | Point |
| Dropoff Centroid Latitude | The latitude of the center of the dropoff census tract or the community area if the census tract has been hidden for privacy. This column often will be blank for locations outside Chicago. | Number |
| Dropoff Centroid Longitude | The longitude of the center of the dropoff census tract or the community area if the census tract has been hidden for privacy. This column often will be blank for locations outside Chicago. | Number |
| Dropoff Centroid Location | The location of the center of the dropoff census tract or the community area if the census tract has been hidden for privacy. This column often will be blank for locations outside Chicago. | Point |


In [3]:
import pandas as pd
import numpy as np
import h3

In [4]:
taxi_data_v1 = pd.read_csv("../data/Taxi_Trips_big.csv")
taxi_data_v1

,Trip ID,Taxi ID,Trip Start Timestamp,Trip End Timestamp,Trip Seconds,Trip Miles,Pickup Census Tract,Dropoff Census Tract,Pickup Community Area,Dropoff Community Area,...,Extras,Trip Total,Payment Type,Company,Pickup Centroid Latitude,Pickup Centroid Longitude,Pickup Centroid Location,Dropoff Centroid Latitude,Dropoff Centroid Longitude,Dropoff Centroid Location
0,7bb35d509b4a23a4d5deae7018a80bdd5f87c952,8b07f9156e568a37d362463c84dbd1118b4eeb753bae50...,06/01/2026 12:00:00 AM,06/01/2026 12:00:00 AM,300.000,"1,2",NaN,NaN,32.0,32.0,...,"$1,00","$11,25",Credit Card,Choice Taxi Association Inc,"41,878865584","-87,625192142",POINT (-87.6251921424 41.8788655841),"41,878865584","-87,625192142",POINT (-87.6251921424 41.8788655841)
1,6630e8872591a9a915ad804f5ce3b34a4654663a,6cf97b9e8e3e1c183c090cdd868df381de772a52987239...,06/01/2026 12:00:00 AM,06/01/2026 12:00:00 AM,445.000,"1,5",NaN,NaN,32.0,28.0,...,"$0,00","$8,92",Mobile,Flash Cab,"41,878865584","-87,625192142",POINT (-87.6251921424 41.8788655841),"41,874005383","-87,66351755",POINT (-87.6635175498 41.874005383)
2,51e2af26f6d9c93afc6e4f1b26a9bf6a3fa7700d,3665a72ee495b03f4dae72307dc6e5e58e21518f77d8e6...,06/01/2026 12:00:00 AM,06/01/2026 12:00:00 AM,0.000,0,NaN,NaN,28.0,28.0,...,"$0,00","$10,38",Credit Card,Transit Administrative Center Inc,"41,874005383","-87,66351755",POINT (-87.6635175498 41.874005383),"41,874005383","-87,66351755",POINT (-87.6635175498 41.874005383)
3,fe861de7a4b804177aa770afd5134dd22758a593,780424fddffb94d4d8541780c7ad1a953ef87d12a20def...,06/01/2026 12:00:00 AM,06/01/2026 12:00:00 AM,4.000,0,NaN,NaN,72.0,72.0,...,"$0,00","$48,60",Credit Card,Globe Taxi,"41,713148612","-87,675075312",POINT (-87.6750753124 41.713148612),"41,713148612","-87,675075312",POINT (-87.6750753124 41.713148612)
4,e2627bcf1c079e0b086cb8dbfc34d6ea955a899a,d5c821e78bf25431862f40219794e4d023679198afeb74...,06/01/2026 12:00:00 AM,06/01/2026 12:00:00 AM,380.000,"1,3",NaN,NaN,32.0,8.0,...,"$0,00","$8,33",Mobile,Taxicab Insurance Agency Llc,"41,878865584","-87,625192142",POINT (-87.6251921424 41.8788655841),"41,899602111","-87,633308037",POINT (-87.6333080367 41.899602111)
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
16086250,fc08277cee8647cb9a4eb700da02ce287029ca2a,3618045f9110d4d88482266ade23659c1a50d32ac37f20...,01/01/2024 12:00:00 AM,01/01/2024 12:15:00 AM,840.000,"7,1",NaN,NaN,41.0,32.0,...,"$0,00","$20,25",Cash,Taxi Affiliation Services,"41,794090253","-87,592310855",POINT (-87.592310855 41.794090253),"41,878865584","-87,625192142",POINT (-87.6251921424 41.8788655841)
16086251,be2ab4c001613a36d473357c2f81c214554ef310,73052f4ccaf4e0fa9178722e491f8e5eda869f56e08aa4...,01/01/2024 12:00:00 AM,01/01/2024 12:45:00 AM,2.820,0,NaN,NaN,28.0,NaN,...,"$0,00","$60,50",Unknown,Taxi Affiliation Services,"41,874005383","-87,66351755",POINT (-87.6635175498 41.874005383),NaN,NaN,NaN
16086252,bdc420394ce5e864465df0a361dfbe95a4e228c4,389f01c14b097ed951468ff163ccc71ebcb99a27e523e9...,01/01/2024 12:00:00 AM,01/01/2024 12:30:00 AM,1.369,"3,07",NaN,NaN,8.0,24.0,...,"$1,00","$26,00",Mobile,Medallion Leasin,"41,899602111","-87,633308037",POINT (-87.6333080367 41.899602111),"41,901206994","-87,676355989",POINT (-87.6763559892 41.9012069941)
16086253,70679bacef2403edfa376cd42f2b04037e7534c1,8123ec5a70681a139fcd93fd477eec0ee05dda82987748...,01/01/2024 12:00:00 AM,01/01/2024 12:15:00 AM,584.000,"0,43",NaN,NaN,8.0,8.0,...,"$1,50","$7,75",Cash,Globe Taxi,"41,899602111","-87,633308037",POINT (-87.6333080367 41.899602111),"41,899602111","-87,633308037",POINT (-87.6333080367 41.899602111)


In [5]:
taxi_data_v1.info()

<class 'pandas.DataFrame'>
RangeIndex: 16086255 entries, 0 to 16086254
Data columns (total 23 columns):
 #   Column                      Dtype  
---  ------                      -----  
 0   Trip ID                     str    
 1   Taxi ID                     str    
 2   Trip Start Timestamp        str    
 3   Trip End Timestamp          str    
 4   Trip Seconds                float64
 5   Trip Miles                  str    
 6   Pickup Census Tract         float64
 7   Dropoff Census Tract        float64
 8   Pickup Community Area       float64
 9   Dropoff Community Area      float64
 10  Fare                        str    
 11  Tips                        str    
 12  Tolls                       str    
 13  Extras                      str    
 14  Trip Total                  str    
 15  Payment Type                str    
 16  Company                     str    
 17  Pickup Centroid Latitude    str    
 18  Pickup Centroid Longitude   str    
 19  Pickup Centroid Location    st

## 1. Delete unneccessary columns
As the features *Fare*, *Tips*, *Tolls*, *Extras* and *Payment Type* are not used for analysis, they are deleted from the dataset.
The features *Pickup Centroid Location* and *Dropoff Centroid Location* are redundant and therefore deleted as well.

In [6]:
cols_to_drop = ["Fare", "Tips", "Tolls", "Extras", "Payment Type", 
                "Pickup Centroid Location", "Dropoff Centroid  Location"]

taxi_data_v1 = taxi_data_v1.drop(columns=cols_to_drop)

## 2. Change Data types from string to numeric/datetime
While downloading the data, some of the data types were changed to *string* and have to be changed back for Analysis.

In [7]:
# Columns to fix data type
cols_to_fix = [
    'Pickup Centroid Longitude', 'Pickup Centroid Latitude',
    'Dropoff Centroid Longitude', 'Dropoff Centroid Latitude',
    'Trip Total', 'Trip Miles'
]

for col in cols_to_fix:
    
    if col == 'Trip Total':
        # Remove the dollar sign and commas from the Trip Total column
        taxi_data_v1[col] = taxi_data_v1[col].astype(str).str.replace('$', '', regex=False).str.replace(',', '', regex=False)
    elif col == 'Trip Miles' or col in ['Pickup Centroid Longitude', 'Pickup Centroid Latitude', 'Dropoff Centroid Longitude', 'Dropoff Centroid Latitude']:
        # Replace the comma with a standard decimal point
        taxi_data_v1[col] = taxi_data_v1[col].astype(str).str.replace(',', '.', regex=False)

    # Convert to numeric
    taxi_data_v1[col] = pd.to_numeric(taxi_data_v1[col], errors='coerce')

#convert to datetime
taxi_data_v1["Trip Start Timestamp"] = pd.to_datetime(
    taxi_data_v1["Trip Start Timestamp"]
)

taxi_data_v1["Trip End Timestamp"] = pd.to_datetime(
    taxi_data_v1["Trip End Timestamp"]
)

/var/folders/mb/d7qxmv150dz1wszhty3l250c0000gn/T/ipykernel_88981/1480342932.py:21: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  taxi_data_v1["Trip Start Timestamp"] = pd.to_datetime(
/var/folders/mb/d7qxmv150dz1wszhty3l250c0000gn/T/ipykernel_88981/1480342932.py:25: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  taxi_data_v1["Trip End Timestamp"] = pd.to_datetime(


In [8]:
taxi_data_v1.info()

<class 'pandas.DataFrame'>
RangeIndex: 16086255 entries, 0 to 16086254
Data columns (total 16 columns):
 #   Column                      Dtype         
---  ------                      -----         
 0   Trip ID                     str           
 1   Taxi ID                     str           
 2   Trip Start Timestamp        datetime64[us]
 3   Trip End Timestamp          datetime64[us]
 4   Trip Seconds                float64       
 5   Trip Miles                  float64       
 6   Pickup Census Tract         float64       
 7   Dropoff Census Tract        float64       
 8   Pickup Community Area       float64       
 9   Dropoff Community Area      float64       
 10  Trip Total                  float64       
 11  Company                     str           
 12  Pickup Centroid Latitude    float64       
 13  Pickup Centroid Longitude   float64       
 14  Dropoff Centroid Latitude   float64       
 15  Dropoff Centroid Longitude  float64       
dtypes: datetime64[us](2), float

## 3. Null Value Analysis
As Null values are difficult to handle during Analysis, they have to be dealt with beforehand.

In [9]:
# Find Null Values and add percentage of NaN values for each column
is_na_df = taxi_data_v1.isna().sum()
is_na_df = pd.DataFrame(is_na_df, columns=['NaN Count'])
is_na_df['Total Count'] = len(taxi_data_v1)
is_na_df['NaN Percentage'] = (is_na_df['NaN Count'] / is_na_df['Total Count']) * 100

# Check for any NaN values in all columns
print("NaN values before conversion:")
print(is_na_df)

NaN values before conversion:
                            NaN Count  Total Count  NaN Percentage
Trip ID                             0     16086255        0.000000
Taxi ID                            12     16086255        0.000075
Trip Start Timestamp                0     16086255        0.000000
Trip End Timestamp                427     16086255        0.002654
Trip Seconds                     3081     16086255        0.019153
Trip Miles                        157     16086255        0.000976
Pickup Census Tract           8891026     16086255       55.270950
Dropoff Census Tract          9102541     16086255       56.585831
Pickup Community Area          445622     16086255        2.770204
Dropoff Community Area        1414559     16086255        8.793588
Trip Total                      32557     16086255        0.202390
Company                             0     16086255        0.000000
Pickup Centroid Latitude       437472     16086255        2.719539
Pickup Centroid Longitude      4

As the columns *Taxi_ID*, *Trip End Timestamp*, *Trip Seconds*, *Trip Miles* and *Trip Total* are needed for the analysis, and the number of null values are relatively small (less than 0.02%), those datapoints are simply deleted from the dataset.

In [10]:
taxi_data_v2 = taxi_data_v1.dropna(subset=['Taxi ID', 'Trip End Timestamp', 
                                                    'Trip Seconds', 'Trip Miles',
                                                    'Trip Total'])
taxi_data_v2.isna().sum()

Trip ID                             0
Taxi ID                             0
Trip Start Timestamp                0
Trip End Timestamp                  0
Trip Seconds                        0
Trip Miles                          0
Pickup Census Tract           8874479
Dropoff Census Tract          9085175
Pickup Community Area          444484
Dropoff Community Area        1410332
Trip Total                          0
Company                             0
Pickup Centroid Latitude       436415
Pickup Centroid Longitude      436415
Dropoff Centroid Latitude     1327009
Dropoff Centroid Longitude    1327009
dtype: int64

For the spatial entities the number of Null values is relatively big, especially for *Census Tract*, so further analysis is neccessary.

In [11]:
spatial_cols = [
    "Pickup Census Tract",
    "Dropoff Census Tract",
    "Pickup Community Area",
    "Dropoff Community Area",
    "Pickup Centroid Latitude",
    "Pickup Centroid Longitude",
    "Dropoff Centroid Latitude",
    "Dropoff Centroid Longitude",
]
null_mask = taxi_data_v2[spatial_cols].isna()

conditional_overlap = pd.DataFrame(
    index=spatial_cols,
    columns=spatial_cols,
    dtype=float
)

for c1 in spatial_cols:
    missing_c1 = null_mask[c1].sum()

    for c2 in spatial_cols:
        if missing_c1 == 0:
            conditional_overlap.loc[c1, c2] = 0
        else:
            conditional_overlap.loc[c1, c2] = (
                (null_mask[c1] & null_mask[c2]).sum() / missing_c1
            )

conditional_overlap = conditional_overlap.round(3)

conditional_overlap

,Pickup Census Tract,Dropoff Census Tract,Pickup Community Area,Dropoff Community Area,Pickup Centroid Latitude,Pickup Centroid Longitude,Dropoff Centroid Latitude,Dropoff Centroid Longitude
Pickup Census Tract,1.000,0.997,0.049,0.121,0.049,0.049,0.121,0.121
Dropoff Census Tract,0.974,1.000,0.045,0.144,0.045,0.045,0.144,0.144
Pickup Community Area,0.977,0.911,1.000,0.671,0.982,0.982,0.655,0.655
Dropoff Community Area,0.759,0.929,0.211,1.000,0.207,0.207,0.941,0.941
Pickup Centroid Latitude,0.995,0.927,1.000,0.668,1.000,1.000,0.667,0.667
Pickup Centroid Longitude,0.995,0.927,1.000,0.668,1.000,1.000,0.667,0.667
Dropoff Centroid Latitude,0.806,0.987,0.220,1.000,0.219,0.219,1.000,1.000
Dropoff Centroid Longitude,0.806,0.987,0.220,1.000,0.219,0.219,1.000,1.000


The overlap matrix shows that data points with missing values in the *Pickup/Dropoff Community Area* fields almost always also have missing values in the corresponding *Census Tract* and *Latitude/Longitude* fields. As these observations do not contain sufficient spatial information for either spatial aggregation, they are removed from the dataset.

More than half of the observations have missing values for the *Census Tract*. This is expected, as Census Tracts are omitted for privacy reasons in some trips and are also unavailable for locations outside the city of Chicago.

Since the *Community Area* and *Latitude/Longitude* values are available for most of these observations, the dataset is split after the data preparation step. For the Census Tract analysis, a dataset containing only observations with available Census Tracts is used. For the Community Area analysis, a larger dataset is retained in which the Census Tract may be missing.

In [10]:
taxi_data_v2

,Trip ID,Taxi ID,Trip Start Timestamp,Trip End Timestamp,Trip Seconds,Trip Miles,Pickup Census Tract,Dropoff Census Tract,Pickup Community Area,Dropoff Community Area,Trip Total,Company,Pickup Centroid Latitude,Pickup Centroid Longitude,Dropoff Centroid Latitude,Dropoff Centroid Longitude
0,7bb35d509b4a23a4d5deae7018a80bdd5f87c952,8b07f9156e568a37d362463c84dbd1118b4eeb753bae50...,2026-06-01,2026-06-01 00:00:00,300.000,1.20,NaN,NaN,32.0,32.0,1125.0,Choice Taxi Association Inc,41.878866,-87.625192,41.878866,-87.625192
1,6630e8872591a9a915ad804f5ce3b34a4654663a,6cf97b9e8e3e1c183c090cdd868df381de772a52987239...,2026-06-01,2026-06-01 00:00:00,445.000,1.50,NaN,NaN,32.0,28.0,892.0,Flash Cab,41.878866,-87.625192,41.874005,-87.663518
2,51e2af26f6d9c93afc6e4f1b26a9bf6a3fa7700d,3665a72ee495b03f4dae72307dc6e5e58e21518f77d8e6...,2026-06-01,2026-06-01 00:00:00,0.000,0.00,NaN,NaN,28.0,28.0,1038.0,Transit Administrative Center Inc,41.874005,-87.663518,41.874005,-87.663518
3,fe861de7a4b804177aa770afd5134dd22758a593,780424fddffb94d4d8541780c7ad1a953ef87d12a20def...,2026-06-01,2026-06-01 00:00:00,4.000,0.00,NaN,NaN,72.0,72.0,4860.0,Globe Taxi,41.713149,-87.675075,41.713149,-87.675075
4,e2627bcf1c079e0b086cb8dbfc34d6ea955a899a,d5c821e78bf25431862f40219794e4d023679198afeb74...,2026-06-01,2026-06-01 00:00:00,380.000,1.30,NaN,NaN,32.0,8.0,833.0,Taxicab Insurance Agency Llc,41.878866,-87.625192,41.899602,-87.633308
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
16086250,fc08277cee8647cb9a4eb700da02ce287029ca2a,3618045f9110d4d88482266ade23659c1a50d32ac37f20...,2024-01-01,2024-01-01 00:15:00,840.000,7.10,NaN,NaN,41.0,32.0,2025.0,Taxi Affiliation Services,41.794090,-87.592311,41.878866,-87.625192
16086251,be2ab4c001613a36d473357c2f81c214554ef310,73052f4ccaf4e0fa9178722e491f8e5eda869f56e08aa4...,2024-01-01,2024-01-01 00:45:00,2.820,0.00,NaN,NaN,28.0,NaN,6050.0,Taxi Affiliation Services,41.874005,-87.663518,NaN,NaN
16086252,bdc420394ce5e864465df0a361dfbe95a4e228c4,389f01c14b097ed951468ff163ccc71ebcb99a27e523e9...,2024-01-01,2024-01-01 00:30:00,1.369,3.07,NaN,NaN,8.0,24.0,2600.0,Medallion Leasin,41.899602,-87.633308,41.901207,-87.676356
16086253,70679bacef2403edfa376cd42f2b04037e7534c1,8123ec5a70681a139fcd93fd477eec0ee05dda82987748...,2024-01-01,2024-01-01 00:15:00,584.000,0.43,NaN,NaN,8.0,8.0,775.0,Globe Taxi,41.899602,-87.633308,41.899602,-87.633308


In [11]:
taxi_data_v3 = taxi_data_v2.dropna(subset=['Pickup Community Area', 'Dropoff Community Area'])
taxi_data_v3.isna().sum()

Trip ID                             0
Taxi ID                             0
Trip Start Timestamp                0
Trip End Timestamp                  0
Trip Seconds                        0
Trip Miles                          0
Pickup Census Tract           7659353
Dropoff Census Tract          7659353
Pickup Community Area               0
Dropoff Community Area              0
Trip Total                          0
Company                             0
Pickup Centroid Latitude            0
Pickup Centroid Longitude           0
Dropoff Centroid Latitude           0
Dropoff Centroid Longitude          0
dtype: int64

## 4. Add H3 Index Columns

For the subsequent spatial analyses, additional spatial entities are added to the dataset by computing H3 indices at resolutions 7, 8, and 9.

The H3 indices are derived from the available pickup and dropoff latitude and longitude coordinates. Due to privacy restrictions, these coordinates do not represent the exact pickup or dropoff locations. Instead, they correspond to the centroid of the associated Census Tract. If the Census Tract is unavailable, the centroid of the corresponding Community Area is provided instead.

Consequently, the H3 indices do not introduce additional spatial information. Instead, they provide an alternative spatial discretization of the city at different levels of granularity, which is used in the subsequent descriptive spatial analysis.

In [12]:
taxi_data_v3['h3_index_pickup_7'] = [
    # Check if lat and lng are not NaN before converting to H3 index, otherwise return None
    h3.latlng_to_cell(lat, lng, 7) if (lat == lat and lng == lng) else None
    for lat, lng in zip(
        taxi_data_v3['Pickup Centroid Latitude'], 
        taxi_data_v3['Pickup Centroid Longitude']
    )
]
taxi_data_v3['h3_index_dropoff_7'] = [
    # Check if lat and lng are not NaN before converting to H3 index, otherwise return None
    h3.latlng_to_cell(lat, lng, 7) if (lat == lat and lng == lng) else None
    for lat, lng in zip(
        taxi_data_v3['Dropoff Centroid Latitude'], 
        taxi_data_v3['Dropoff Centroid Longitude']
    )
]

taxi_data_v3['h3_index_pickup_8'] = [
    # Check if lat and lng are not NaN before converting to H3 index, otherwise return None
    h3.latlng_to_cell(lat, lng, 8) if (lat == lat and lng == lng) else None
    for lat, lng in zip(
        taxi_data_v3['Pickup Centroid Latitude'], 
        taxi_data_v3['Pickup Centroid Longitude']
    )
]
taxi_data_v3['h3_index_dropoff_8'] = [
    # Check if lat and lng are not NaN before converting to H3 index, otherwise return None
    h3.latlng_to_cell(lat, lng, 8) if (lat == lat and lng == lng) else None
    for lat, lng in zip(
        taxi_data_v3['Dropoff Centroid Latitude'], 
        taxi_data_v3['Dropoff Centroid Longitude']
    )
]

taxi_data_v3['h3_index_pickup_9'] = [
    # Check if lat and lng are not NaN before converting to H3 index, otherwise return None
    h3.latlng_to_cell(lat, lng, 9) if (lat == lat and lng == lng) else None
    for lat, lng in zip(
        taxi_data_v3['Pickup Centroid Latitude'], 
        taxi_data_v3['Pickup Centroid Longitude']
    )
]
taxi_data_v3['h3_index_dropoff_9'] = [
    # Check if lat and lng are not NaN before converting to H3 index, otherwise return None
    h3.latlng_to_cell(lat, lng, 9) if (lat == lat and lng == lng) else None
    for lat, lng in zip(
        taxi_data_v3['Dropoff Centroid Latitude'], 
        taxi_data_v3['Dropoff Centroid Longitude']
    )
]

In [13]:
# Check for any NaN values in the new H3 index columns
print(taxi_data_v3[['h3_index_pickup_7', 
                 'h3_index_dropoff_7',
                 'h3_index_pickup_8',
                 'h3_index_dropoff_8',
                 'h3_index_pickup_9',
                 'h3_index_dropoff_9',
                 'Pickup Census Tract', 
                 'Dropoff Census Tract']].isna().sum())

columns_to_drop = [
    'Pickup Centroid Longitude', 'Pickup Centroid Latitude',
    'Dropoff Centroid Longitude', 'Dropoff Centroid Latitude'
]

h3_index_pickup_7             0
h3_index_dropoff_7            0
h3_index_pickup_8             0
h3_index_dropoff_8            0
h3_index_pickup_9             0
h3_index_dropoff_9            0
Pickup Census Tract     7659353
Dropoff Census Tract    7659353
dtype: int64


## 5. Outlier Analysis and Data Quality Checks

The numerical variables are inspected for implausible values. Negative trip durations, trip distances and trip costs would indicate data quality issues and are therefore checked.

In [14]:
taxi_data_v3 = taxi_data_v3[
    (taxi_data_v3["Trip Seconds"] > 0) &
    (taxi_data_v3["Trip Miles"] > 0) &
    (taxi_data_v3["Trip Total"] >= 0)
]

In [15]:
short_trips = taxi_data_v3[taxi_data_v3["Trip Seconds"] < 10]

print(f"Trips with Trip Seconds < 10: {len(short_trips):,}")
print(f"Percentage: {len(short_trips)/len(taxi_data_v3)*100:.2f}%")

comparison = (
    taxi_data_v3
    .assign(short_trip=taxi_data_v3["Trip Seconds"] < 10)
    .groupby("short_trip")
    .agg(
        trips=("Trip ID", "count"),
        median_price=("Trip Total", "median"),
        median_distance=("Trip Miles", "median"),
        same_coordinates=(
            "Trip ID",
            lambda x: (
                (
                    taxi_data_v3.loc[x.index, "Pickup Centroid Latitude"]
                    == taxi_data_v3.loc[x.index, "Dropoff Centroid Latitude"]
                )
                &
                (
                    taxi_data_v3.loc[x.index, "Pickup Centroid Longitude"]
                    == taxi_data_v3.loc[x.index, "Dropoff Centroid Longitude"]
                )
            ).mean()
        )
    )
)

comparison["same_coordinates"] *= 100

comparison

Trips with Trip Seconds < 10: 5,932,052
Percentage: 45.81%


,trips,median_price,median_distance,same_coordinates
short_trip,,,,
False,7017703,1050.0,1.42,14.593992
True,5932052,3775.0,11.97,2.320175


A substantial number of observations report trip durations shorter than 10 seconds. Such travel times are implausible given the corresponding trip distances and therefore indicate a clear data quality issue affecting the Trip Seconds variable. However, these observations do not resemble incomplete or cancelled trips. Compared with the remaining dataset, they exhibit substantially longer travel distances, higher fares, and almost always different pickup and dropoff locations (only 2.3% share identical published coordinates, compared with 14.6% for the remaining observations). These characteristics suggest that the affected records likely represent genuine taxi trips, while the recorded travel time is erroneous. Consequently, the observations are retained in the dataset, whereas analyses relying on trip duration or derived speed metrics treat these records with caution.

In [16]:
weather_end = pd.Timestamp("2026-05-12 23:00:00")
taxi_data_v3 = taxi_data_v3.loc[taxi_data_v3["Trip Start Timestamp"] <= weather_end].copy()

The final day in the dataset contains only a small fraction of the expected number of trips, indicating that the observation period ends before the day is complete. To avoid biased daily and weekly demand statistics, this incomplete day is removed from all subsequent analyses. As the weather data is only available until **2026-05-12 23:00**, taxi trips occurring after this timestamp are removed for subsequent analyses and predictive modeling.

No additional filtering based on statistical outliers is performed druing data preparation, as extreme values may correspond to legitimate taxi trips (e.g., airport transfers or unusually long trips) and therefore contain relevant information for the subsequent analyses.

## 6. Save the dataset as parquet for all further Analysis
The big dataset *taxi_data_processed_big* is being used for Communinty Area Analysis.

In [16]:
taxi_data_v3.to_parquet(
    "../data/processed/taxi_data_processed_big.parquet"
)

The small dataset *taxi_data_processed_small* is being used for Census Tract Analysis.

In [17]:
taxi_data_small = taxi_data_v3.dropna(subset=['Pickup Census Tract', 'Dropoff Census Tract'])

In [18]:
taxi_data_small.to_parquet(
    "../data/processed/taxi_data_processed_small.parquet"
)